File: lintcode/cpp/Rotate_String.cpp
class Solution {
public:
  /**
     * param A: A string
     * param offset: Rotate string with offset.
     * return: Rotated string.
     */
    string rotateString(string A, int offset) {
        // wirte your code here
        if(A.size() == 0) return "";
        
        int n = A.size();
        offset %= n;
        
        int i = n-1;
        int x = A[i];
        int end = n-1;
        int cnt = 1;
        
        while(cnt < n || i != end){
            if(i == end){
                i--;
                x = A[i];
                end = i;
            }
            
            int j = (i-offset+n)%n;
            
            if(j == end) A[i] = x;
            else swap(A[i], A[j]);
            
            i = j;
            cnt++;
        }
        
        return A;
    }
};

File: lintcode/java/RotateString.java
public class Solution {
    /**
     * @param str: an array of char
     * @param offset: an integer
     * @return: nothing
     */
    public void rotateString(char[] str, int offset) {
        if(offset < 0) throw new IllegalArgumentException("xxx");
        if(str.length == 0) return;
        
        offset %= str.length;
        if(offset == 0) return;
        
        int i = 0;
        int j = str.length-1;
        
        reverse(str, i, j);
        reverse(str, i, offset-1);
        reverse(str, offset, j);
    }
    
    private void reverse(char[] A, int i, int j){
        if(i > j) throw new IllegalArgumentException("xxx");
        
        while(i < j){
            swap(A, i++, j--);
        }
    }
    
    private void swap(char[] A, int i, int j){
        char t = A[i];
        A[i] = A[j];
        A[j] = t;
    }
}
